src/greynirseq/nicenlp/data/parallel_documents/indexed_parallel_documents_dataset.py

```diff
@@ -0,0 +1,1015 @@
+        if KEYS.TARGET_OFFSETS in ex_batched:
+            tgt_doc_offsets = ex_batched[KEYS.TARGET_OFFSETS]
+            all_tgt_idxs = [
+                [idx + document_offset for idx in idxs]
+                for (idxs, document_offset) in zip(all_tgt_idxs, tgt_doc_offsets)
+            ]
+
+        # cast for readability
+        all_src_idxs = cast(List[List[int]], all_src_idxs)
+        all_tgt_idxs = cast(List[List[int]], all_tgt_idxs)
+        doc_idxs = cast(List[int], doc_idxs)
+        pg_idxs = cast(List[int], pg_idxs)
+        weights = cast(List[int], weights)
+        skip = cast(List[bool], skip)
+        abs_align_indices = cast(List[int], abs_align_indices)
+
+        ex_out = {
+            KEYS.SOURCE_INDICES: [],
+            KEYS.TARGET_INDICES: [],
+            KEYS.WEIGHT: [],
+            KEYS.EXACT_ALIGNMENT: [],
+        }
+        # set up reproducible rng state that depends implicitly on batch_size but is invariant to num_proc
+        rng = np.random.default_rng((seed, abs_align_indices))
+        # these are the sequence lengths of the bins we will use to merge sentences
+        extra_padding = max_merges + 1 + 1 + 1  # SENT_SEP*max_merges + Start + End + Maybe BT info
+        maximum_lengths = np.array([50, 100, 150, 350, max_seq_len - extra_padding], dtype=np.int64)
+
+        # fetch maximum number of rolls to minimize fn calls
+        bin_idxs = np.clip(rng.poisson(_POISSON_MEAN, size=len(doc_idxs)), 0, len(maximum_lengths) - 1)
+        # examples to pass through without merging
+        ok_to_merge = rng.random(len(doc_idxs)) > no_merge_prob
+
+        # Initial conditions
+        last_context_idxs: Optional[Tuple[int, int]] = None
+        accum_src, accum_tgt, accum_weight = (
+            [],
+            [],
+            0,
+        )
+        num_outputs, curr_num_merges = 0, 0
+        for _loop_idx, (
+            doc_idx,
+            pg_idx,
+            weight,
+            src_idxs,
+            tgt_idxs,
+            skip_example,
+        ) in enumerate(zip(doc_idxs, pg_idxs, weights, all_src_idxs, all_tgt_idxs, skip)):
+            # This loop is a bit tricky to understand, here is a high level overview:
+            # We iterate over the examples in the batch, and for each example we roll a dice to decide whether to
+            # merge it with the next example or not. If we decide to merge, we check whether the accumulated weight
+            # is within the maximum sequence length, if it is, we merge the current example with the accumulator
+            # and continue. If it is not, we store the accumulator and reset it to the current example.
+            # An implicit assumption here is that the examples have not been shuffled,
+            # so that nearby examples "are in context"
+
+            # examples are skipped if they are structurally invalid: too long, too short (no fragments) or src and tgt
+            # have very different relative lengths
+            if skip_example:
```

Comment:
Do we want to merge over skipped examples or should we reset the accumulator here?

The same question applies to the if statement in 964.